content/articles/django-custom-template-tags-and-filters/index.md

```diff
@@ -0,0 +1,258 @@
+│   │
+│   ├── tests.py
+│   ├── url.py
+│   └── views.py
+├── Contact
+│   ├── asgi.py
+│   ├── settings.py
+│   ├── urls.py
+│   └── wsgi.py
+├── db.sqlite3
+└──  manage.py
+```
+
+You can create as many modules as you can inside the `templatetags` folder. You should, however, note that using the `{% load %}` tag will only work for a particular module name, not the app name.
+
+```python
+from Django import template
+register = template.Library()
+```
+
+To load the tags and filters, we will do the following in our template:
+
+```jinja
+{% load custom_tags %}
+```
+
+### Writing Custom Filters
+Filters are usually written either as `{{value|filter}}` or `{{value|filter:'arg'}}` . This means that the filter will or will not take a argument when acting on a value.
+
+Consider a simple contacts list application showing a contact and when they were created.
+
+The model.py looks as follows:
+
+```Python
+from Django.db import models
+
+class Contact(models.Model):
+    name = models.CharField(max_length=50)
+    date = models.DateField()
+```
+
+The application looks as follows with just a little of styling...
+
+![Contact App](/engineering-education/Django-custom-template-tags-and-filters/image1.png)
+
+Now let's create two simple filters that capitalize all strings in a given name, the first character of a word and also sets a color.
+
+We use `@register` to register our filter and give it a name in Django. Our function takes a name as it's value.
+
+To use our filter we use the line `{% load custom_tags %}` in our templates.
+
+template_tags/custom_tags.py
+
+```python
+from Django import template
+
+register = template.Library()
+
+@register.filter(name='upper')
+def upper(value):
+  return value.upper()
+
+@register.filter(name='modify_name')
+def modify_name(value):
+    return value.title()
+
+@register.filter(name='get_color')
+def color(value):
+    if value:
+        return "#FF0000"
+```
+
+contact.html
+
+```html
+{% load custom_tags %}
+
+
+{% block content %}
+
+{% for contact in contacts %}
+<div class="flex-box-container">
+    <div style="background-color: {{contact.name|get_color}};"><p >{{contact.name|upper}}</p></div>
+    <div><b>Created:</b> {{ contact.date }}</div>
+</div>
+{% endfor %}
+
+{% endblock content %}
+```
+
+The resulting view will look as shown below:
+
+![Contact App](/engineering-education/Django-custom-template-tags-and-filters/image2.png)
+
+The `upper ` filter capitalizes the contact name while the `get_color ` filter sets a new color to the div.
+
+Likewise, you can create your own custom filters to cater for whatever needs you may have for your templates.
+
+You should consider the names supplied to your views. They should match those used with `@register.filter()`. For instance, using `color` instead of `get_color` would result to an error. You may also have to consider the auto-escaping behaviour of Django with the filter. You may look up on Django documentation to learn more.
+
```

Comment:
Provide a link to the section in Django docs that talks about auto-escaping.

Reply:
Done